stdlib/public/SDK/Foundation/DataProtocol.swift

```diff
@@ -0,0 +1,239 @@
+    @discardableResult
+    func copyBytes<R: RangeExpression>(to: UnsafeMutableRawBufferPointer, from: R) -> Int where R.Bound == Index
+    @discardableResult
+    func copyBytes<DestinationType, R: RangeExpression>(to: UnsafeMutableBufferPointer<DestinationType>, from: R) -> Int where R.Bound == Index
+    
+    
+    var _borrowedBytes: (buffer: UnsafeRawBufferPointer, owner: Any)? { get }
+}
+
+public protocol MutableDataProtocol : DataProtocol, MutableCollection, RangeReplaceableCollection {
+    /// Replaces the contents of the buffer at the given range with zeroes.
+    /// Default implementation can be given in terms of `replaceSubrange(_:with:)`
+    mutating func resetBytes<R: RangeExpression>(in range: R) where R.Bound == Index
+}
+
+// -- DataProtocol Extensions --------------------------------------------------
+extension DataProtocol {    
+    // -- Searching ------------------------------------------------------------
+    public func firstRange<D: DataProtocol>(of data: D) -> Range<Index>? {
+        return self.firstRange(of: data, in: self.startIndex ..< self.endIndex)
+    }
+    
+    public func lastRange<D: DataProtocol>(of data: D) -> Range<Index>? {
+        return self.firstRange(of: data, in: self.startIndex ..< self.endIndex)
+    }
+    
+    // -- Copying --------------------------------------------------------------
+    @discardableResult
+    public func copyBytes(to ptr: UnsafeMutableRawBufferPointer) -> Int {
+        return copyBytes(to: ptr, from: self.startIndex ..< self.endIndex)
+    }
+
+    @discardableResult
+    public func copyBytes<DestinationType>(to ptr: UnsafeMutableBufferPointer<DestinationType>) -> Int {
+        return copyBytes(to: ptr, from: self.startIndex ..< self.endIndex)
+    }
+    
+    @discardableResult
+    public func copyBytes(to ptr: UnsafeMutableRawBufferPointer, count: Int) -> Int {
+        return copyBytes(to: ptr, from: self.startIndex ..< self.index(self.startIndex, offsetBy: count))
+    }
+    
+    @discardableResult
+    public func copyBytes<DestinationType>(to ptr: UnsafeMutableBufferPointer<DestinationType>, count: Int) -> Int {
+        return copyBytes(to: ptr, from: self.startIndex ..< self.index(self.startIndex, offsetBy: count))
+    }
+    
+    @discardableResult
+    public func copyBytes<R: RangeExpression>(to ptr: UnsafeMutableRawBufferPointer, from range: R) -> Int where R.Bound == Index {
+        let bound = ptr.bindMemory(to: UInt8.self)
+        return copyBytes(to: bound, from: range)
+    }
+    
+    @discardableResult
+    public func copyBytes<DestinationType, R: RangeExpression>(to ptr: UnsafeMutableBufferPointer<DestinationType>, from range: R) -> Int where R.Bound == Index {
+        precondition(ptr.baseAddress != nil)
+
+        let concreteRange = range.relative(to: self)
+        let slice = self[concreteRange]
+
+        return ptr.withMemoryRebound(to: UInt8.self) { buffer -> Int in
```

Comment:
In general, binding or rebinding the memory type is a huge red flag--I
only expect to see it a workaround for some legacy API.  In this case,
you want to view a typed buffer as a sequence of bytes, so
`ptr.withUnsafeBytes()` is what you want. Why won't that work?

Reply:
Thanks for reviewing — we want to ensure we're doing 100% the right thing here. For the sake of understanding: how does `ptr.withUnsafeBytes()` differ from temporarily rebinding to `UInt8` from a memory safety perspective?